_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,295 @@
+
+
+## Hybrid search using RRF
+
+Rank-based techniques focus on a document's position in the result set, rather than its raw score. These techniques are especially useful when combining results from different retrieval methods whose scores aren't directly comparable. 
+
+**RRF** is one such technique. It uses the rank of each document in individual query results to calculate a combined score, making it robust against mismatched scoring scales across methods like BM25 and semantic search.
+
+Define the RRF search pipeline:
+
+```json
+PUT /_search/pipeline/rrf-pipeline
+{
+  "description": "Post processor for hybrid RRF search",
+  "phase_results_processors": [
+    {
+      "score-ranker-processor": {
+        "combination": {
+          "technique": "rrf"
+        }
+      }
+    }
+  ]
+}
+```
+
+Run the hybrid query:
+
+```json
+GET /bookstore_catalog/_search?search_pipeline=rrf-pipeline
+{
+  "query": {
+    "hybrid": {
+      "queries": [
+        {
+          "match": {
+            "title": {
+              "query": "science fiction"
+            }
+          }
+        },
+        {
+          "neural": {
+            "title_embedding": {
+              "query_text": "science fiction",
+              "model_id": "aVeif4oB5Vm0Tdw8zYO2",
+              "k": 5
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+This query uses:
+
+- `hybrid`: Combines results from lexical and semantic search.
+- `search_pipeline`: Applies the RRF combination logic to rank the final results.
+
+### How RRF works
+
+RRF works as follows:
+
+1. **Sort documents by score**: Each query method sorts documents by score.
+1. **Assign rank positions**: Documents are ranked based on their scores for each query.
+1. **Compute the RRF score**: For each document, the RRF score is computed using the following formula:
+    ```python
+    rankScore(document_i) = sum((1/(k + query_1_rank), (1/(k + query_2_rank), ..., (1/(k + query_j_rank)))
+    ```
+
+    where:
+    -  `k` is a rank constant 
+    - `query_j_rank` represents the ranking of a document in a particular query method. 
+
+1. **Add rank contributions**: Rank calculations are combined, and documents are sorted by decreasing rank score.
+1. **Return the top results**: The highest-ranked documents are retrieved based on the query size.
+
+## When to use rank-based normalization
+
+Consider using rank-based normalization in the following situations:
+
```

Comment:
Line 275: "by only using result ranking" ("ranking" instead of "rank")? Please use a word other than "dominate" and provide context. Dominate what?

Reply:
Dominate the score. Reworded.